src/js/common/schemaform/SchemaForm.jsx

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import _ from 'lodash/fp';
+import Form from 'react-jsonschema-form';
+
+import { uiSchemaValidate, transformErrors } from './validation';
+import Address from './Address';
+import FieldTemplate from './FieldTemplate';
+import * as reviewWidgets from './review/widgets';
+import ReviewFieldTemplate from './review/ReviewFieldTemplate';
+import StringField from './review/StringField';
+import widgets from './widgets/index';
+import ObjectField from './ObjectField';
+import ArrayField from './ArrayField';
+import ReviewObjectField from './review/ObjectField';
+import { scrollToFirstError } from '../utils/helpers';
+
+const fields = {
+  ObjectField,
+  ArrayField,
+  address: Address
+};
+
+const reviewFields = {
+  ObjectField: ReviewObjectField,
+  ArrayField,
+  address: ReviewObjectField,
+  StringField
+};
+
+/*
+ * Each page uses this component and passes in config. This is where most of the page level
+ * form logic should live.
+ */
+class SchemaForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.validate = this.validate.bind(this);
+    this.onError = this.onError.bind(this);
+    this.getEmptyState = this.getEmptyState.bind(this);
+    this.transformErrors = this.transformErrors.bind(this);
+    this.state = this.getEmptyState(props);
+  }
+
+  componentWillReceiveProps(newProps) {
+    if (newProps.name !== this.props.name) {
+      this.setState(this.getEmptyState(newProps));
+    }
+  }
+
+  onError() {
+    const formContext = _.set('submitted', true, this.state.formContext);
+    this.setState({ formContext });
+    scrollToFirstError();
+  }
+
+  getEmptyState() {
+    const { onEdit, hideTitle, title } = this.props;
+    return {
+      formContext: {
+        touched: {},
+        submitted: false,
+        onEdit,
+        hideTitle,
+        pageTitle: title
+      }
+    };
+  }
+
+  transformErrors(errors) {
+    return transformErrors(errors, this.props.uiSchema);
```

Comment:
This is taking the list of errors from JSON Schema and doing two things:

1. It moves required errors to the field instead of the object level. This is a quirk of JSON Schema that missing field errors are associated with the containing object, instead of the field itself. So like if you're missing a last name in our `veteranFullName` field, JSON Schema marks `veteranFullName` as having an error instead of `veteranFullName.last`. We change that so the error message shows up in the right place.
2. It replaces the JSON Schema error messages with messages from uiSchema or default ones in validation.js. The default error messages are pretty bad.